tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -386,188 +398,956 @@ impl ServerKey {
+
+                rayon::join(
+                    || {
+                        let block = output_flag.as_mut().unwrap();
+                        self.key.unchecked_add_assign(
+                            block,
+                            &resolved_carries[resolved_carries.len() - 1],
+                        );
+                        self.key.apply_lookup_table_assign(block, &carry_flag_lut);
+                    },
+                    add_carries_and_cleanup,
+                );
+            }
+        }
+
+        blocks.clone_from_slice(&prepared_blocks);
+
+        match requested_flag {
+            OutputFlag::None => None,
+            OutputFlag::Overflow | OutputFlag::Carry => {
+                output_flag.map(BooleanBlock::new_unchecked)
+            }
+        }
     }
 
-    /// Backbone algorithm of parallel carry (only one bit) propagation
-    ///
-    /// Uses the Hillis and Steele prefix scan
-    ///
-    /// Requires the blocks to have at least 4 bits
-    pub(crate) fn compute_carry_propagation_parallelized_low_latency(
+    fn compute_shifted_blocks_and_block_states(
         &self,
-        generates_or_propagates: Vec<Ciphertext>,
-    ) -> (Vec<Ciphertext>, Ciphertext) {
-        if generates_or_propagates.is_empty() {
-            return (vec![], self.key.create_trivial(0));
-        }
+        blocks: &[Ciphertext],
+    ) -> (Vec<Ciphertext>, Vec<Ciphertext>) {
+        let num_blocks = blocks.len();
 
-        let lut_carry_propagation_sum = self
-            .key
-            .generate_lookup_table_bivariate(prefix_sum_carry_propagation);
-        // Type annotations are required, otherwise we get confusing errors
-        // "implementation of `FnOnce` is not general enough"
-        let sum_function = |block_carry: &mut Ciphertext, previous_block_carry: &Ciphertext| {
-            self.key.unchecked_apply_lookup_table_bivariate_assign(
-                block_carry,
-                previous_block_carry,
-                &lut_carry_propagation_sum,
+        let message_modulus = self.message_modulus().0 as u64;
+
+        let block_modulus = self.message_modulus().0 * self.carry_modulus().0;
+        let num_bits_in_block = block_modulus.ilog2();
+
+        let grouping_size = num_bits_in_block as usize;
+
+        let shift_block_fn = |block| (block % message_modulus) << 1;
+        let mut first_grouping_luts = vec![{
+            let first_block_state_fn = |block| {
+                if block >= message_modulus {
+                    1 // Generates
+                } else {
+                    0 // Nothing
+                }
+            };
+            self.key
+                .generate_many_lookup_table(&[&first_block_state_fn, &shift_block_fn])
+        }];
+        for i in 1..grouping_size {
+            let state_fn = |block| {
+                let r = if block >= message_modulus {
+                    2 // Generates Carry
+                } else if block == message_modulus - 1 {
```

Comment:
IIRC from your sprint demo this is one of the genius part of the new algo, a small illustration might be useful to show how the addition of the bits will naturally compute the propagation